### PR TITLE
Small artifacts for bloat reports

### PR DESCRIPTION
**Pull request by kpschoedel**
Small-artifact bloat reports

#### Problem

The current bloat report requires preserving large binaries, and has
trouble matching tree parents for comparison.

#### Change overview

This change makes example builds generate small artifacts containing
json file(s) containing build item identification and section sizes.

Artifacts names have the form “Size,PR,WORKFLOW,CURRENT_SHA,PARENT_SHA”;
these contain one or more json files, each containing a report on one
build target.

The reporting scripts then examines the list of available artifacts to
list of artifacts to identify pairs with the same $WORKFLOW where one's
CURRENT_SHA is the other's PARENT_SHA, and only then downloads and
processes those artifacts to generate comparisons for matching builds.

- Changes to `examples-…` workflows:

  - Adds some `GH_EVENT_…` variables to workflow `env`, to identify
    the current run.
  - Uses the `gh_sizes.py` script to generate size report json files.
  - Uploads size report artifacts.

- Changes to `scripts/tools/memory`:
  - Adds minimal platform config files for recently-added platforms.
  - Adds markdown output options (required for github comments) along
    with some associated cleanup.
  - Adds a script `scripts/tools/memory/gh_sizes.py` for use by workflows;
    this is similar to `report_summary.py` with a suitable consistent set
    of arguments.
  - Adds a script `scripts/tools/memory/gh_report.py` to analyze size
    report artifacts.

- Modifies `bloat_check.py` to ignore the size report artifacts.

The github comments produced by `gh_report.py` are slightly different
fromt the existing reports. Since this change enables reports for many
more builds, a full report will be over a hundred lines, and is placed
inside a details tag. Only size changes above a configurable threshold
are called out ‘above the fold’.

Note that this PR does _not_ include a change to actually send size
report comments on github; this is left to a followup after that code
has been verified on real-world size artifacts.

#### Testing

Fork-CI runs and offline verification on resulting artifacts.

### CODE REVIEW

**Comment by kpschoedel** [comment]
> chip-lock-app.elf	.flash.text	-68	-68
> chip-bridge-app.elf	.flash.text	-48	-48
> chip-ipv6only-app.elf	.flash.text	172	172
> chip-shell.elf	.flash.text	-56	-56

_Didn't actually change any code_

**Comment by kpschoedel** [comment]
Concrete example: [This EFR32 build of this PR](https://github.com/project-chip/connectedhomeip/pull/9331/checks?check_run_id=3463730299) produces the artifact [Sizes,EFR32-Examples,ae4a3faed3499547551a558d60a438555b8a876c,cb5e3959105b35358b9b594bad1cb3a8328d9150](https://github.com/project-chip/connectedhomeip/suites/3632720232/artifacts/87951750) which contains four json files, one for each binary; `chip-efr32-lock-example.out-sizes.json` contains:
```{json}
{
  "platform": "efr32",
  "config": "BRD4161A",
  "build": "lock-example",
  "event": "pull_request",
  "hash": "ae4a3faed3499547551a558d60a438555b8a876c",
  "parent": "cb5e3959105b35358b9b594bad1cb3a8328d9150",
  "pr": "9331",
  "input": "out/lock_app_debug/BRD4161A/chip-efr32-lock-example.out",
  "time": "1630339790",
  "frames": {
    "sizes": [{"section":".bss","size":260496},{"section":".data","size":1648},{"section":".text","size":741772}]}}
```

The report generator will look for a corresponding artifact named `Sizes,EFR32-Examples,cb5e3959105b35358b9b594bad1cb3a8328d9150,*` and produce a comparison for summaries with matching (`platform`,`config`,`build`).

**Comment by kpschoedel** [comment]
Sigh, need to add numpy/pandas to the build containers.

**Review by msandstedt — APPROVED** [review]

**Comment by kpschoedel** [comment]
Note: the bloat check workflow is expected to run this with something like :
```
scripts/tools/memory/gh_report.py \
  --verbose \
  --report-increases 1 \
  --report-pr \
  --github-comment \
  --github-repository project-chip/connectedhomeip \
  --github-api-token "${{ secrets.GITHUB_TOKEN }}"
```
(in this example highlighting size increases over 1%)

**Comment by mspang** [comment]
Do you have an example of what this looks like?

File: scripts/tools/memory/platform/linux.cfg [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+                ".strtab",
+                ".symtab",
+                ".text",
+            ],
+            'RAM': [
+                ".bss",
+                ".ctors",
+                ".data1",
+                ".data.rel.ro",
+                ".data",
```
Comment (mspang) [inline scripts/tools/memory/platform/linux.cfg:64]: This also goes into flash right? Should it be listed there as well?

Comment (mspang) [reply scripts/tools/memory/platform/linux.cfg:64]: Honestly shouldn't we use the program headers to determine the file size and ram size?

You can't tell from the name...

**Comment by kpschoedel** [comment]
> Do you have an example of the reports will look like?

A small fake-data example is below. For actual runs, the table would contain all the available builds. I'll try to add a fuller example later.

---

PR #1234: Size comparison from d571fb99094d7f5a713879df6a6496ab10f1dd8c to e1ea9109a6853b5a36c8f1fd85914aedf05aa0f7

**Increases above 1.0% from d571fb99094d7f5a713879df6a6496ab10f1dd8c to e1ea9109a6853b5a36c8f1fd85914aedf05aa0f7:**

| platform   | target       | config                      | section     |   e1ea9109 |   d571fb99 |   change |   % change |
|:-----------|:-------------|:----------------------------|:------------|-----------:|-----------:|---------:|-----------:|
| mbed       | lighting-app | CY8CPROTO_062_4343W+release | .data       |       4960 |       5016 |       56 |        1.1 |
|            |              |                             | .text       |    1087392 |    1109600 |    22208 |        2.0 |
|            |              |                             | .zero.table |          0 |          8 |        8 |      inf   |

<details>
<summary>1 build</summary>

| platform   | target       | config                      | section       |   e1ea9109 |   d571fb99 |   change |   % change |
|:-----------|:-------------|:----------------------------|:--------------|-----------:|-----------:|---------:|-----------:|
| mbed       | lighting-app | CY8CPROTO_062_4343W+release | *ABS*         |       4096 |       4096 |        0 |        0.0 |
|            |              |                             | .bss          |     170044 |     171396 |     1352 |        0.8 |
|            |              |                             | .copy.table   |         24 |         24 |        0 |        0.0 |
|            |              |                             | .cy_m0p_image |       5996 |       5996 |        0 |        0.0 |
|            |              |                             | .cy_sharedmem |          8 |          8 |        0 |        0.0 |
|            |              |                             | .data         |       4960 |       5016 |       56 |        1.1 |
|            |              |                             | .heap         |     862464 |     861056 |    -1408 |       -0.2 |
|            |              |                             | .noinit       |        120 |        120 |        0 |        0.0 |
|            |              |                             | .ramVectors   |        736 |        736 |        0 |        0.0 |
|            |              |                             | .text         |    1087392 |    1109600 |    22208 |        2.0 |
|            |              |                             | .zero.table   |          0 |          8 |        8 |      inf   |

</details>

**Comment by mspang** [comment]
> > Do you have an example of the reports will look like?
> 
> A small fake-data example is below. For actual runs, the table would contain all the available builds. I'll try to add a fuller example later.
> 
> PR #1234: Size comparison from [d571fb9](https://github.com/project-chip/connectedhomeip/commit/d571fb99094d7f5a713879df6a6496ab10f1dd8c) to [e1ea910](https://github.com/project-chip/connectedhomeip/commit/e1ea9109a6853b5a36c8f1fd85914aedf05aa0f7)
> 
> **Increases above 1.0% from [d571fb9](https://github.com/project-chip/connectedhomeip/commit/d571fb99094d7f5a713879df6a6496ab10f1dd8c) to [e1ea910](https://github.com/project-chip/connectedhomeip/commit/e1ea9109a6853b5a36c8f1fd85914aedf05aa0f7):**
> 
> platform	target	config	section	[e1ea910](https://github.com/project-chip/connectedhomeip/commit/e1ea9109a6853b5a36c8f1fd85914aedf05aa0f7)	[d571fb9](https://github.com/project-chip/connectedhomeip/commit/d571fb99094d7f5a713879df6a6496ab10f1dd8c)	change	% change
> mbed	lighting-app	CY8CPROTO_062_4343W+release	.data	4960	5016	56	1.1
> .text	1087392	1109600	22208	2.0
> .zero.table	0	8	8	inf
> 1 build

Nice.. I've really missed the baseline in the current reports.

**Review by mspang — APPROVED** [review]

File: scripts/tools/memory/platform/linux.cfg (hunk 1 above)
Comment (kpschoedel) [reply scripts/tools/memory/platform/linux.cfg:64]: > This also goes into flash right? Should it be listed there as well?

It's marked SHF_WRITE, so I put it under ‘RAM’.

> Honestly shouldn't we use the program headers to determine the file size and ram size?

Yes. This is actually a slightly different facility, for arbitrarily grouping sections (from a previous round of memory investigations, not new for this PR.) An actual segment report would be good, but I haven't gotten to that yet.

**Comment by kpschoedel** [comment]
Here's an example of what a full report would look like, given the workflows in this PR. (This is a comparison of two handy builds on my fork, and doesn't match an actual commit or PR.)

---

Size comparison from a184f57ccb1e1513e9ef322b174360b2f138931a to df7771f3b53d80be5fbbdb3802064bcf39bd3df8

**Increases above 1.0% from a184f57ccb1e1513e9ef322b174360b2f138931a to df7771f3b53d80be5fbbdb3802064bcf39bd3df8:**

| platform   | target                 | config        | section   |   df7771f3 |   a184f57c |   change |   % change |
|:-----------|:-----------------------|:--------------|:----------|-----------:|-----------:|---------:|-----------:|
| k32w       | shell                  | k32w061+debug | .data     |        660 |        672 |       12 |        1.8 |
| linux      | ota-provider-app       | debug         | RAM       |      60792 |      61520 |      728 |        1.2 |
|            | shell                  | debug         | RAM       |      55832 |      56576 |      744 |        1.3 |
| qpg        | lighting-app           | qpg6100+debug | .data     |        980 |        992 |       12 |        1.2 |
|            | lock-app               | qpg6100+debug | .data     |        916 |        928 |       12 |        1.3 |
|            | persistent-storage-app | qpg6100+debug | .data     |        268 |        288 |       20 |        7.5 |
|            |                        |               | .text     |     107092 |     108788 |     1696 |        1.6 |

<details>
<summary>40 builds</summary>

| platform   | target                      | config                      | section       |   df7771f3 |   a184f57c |   change |   % change |
|:-----------|:----------------------------|:----------------------------|:--------------|-----------:|-----------:|---------:|-----------:|
| efr32      | lighting-app                | BRD4161A                    | .bss          |     260360 |     260352 |       -8 |       -0.0 |
|            |                             |                             | .data         |       1780 |       1792 |       12 |        0.7 |
|            |                             |                             | .text         |     752612 |     753364 |      752 |        0.1 |
|            | lock-app                    | BRD4161A                    | .bss          |     260424 |     260408 |      -16 |       -0.0 |
|            |                             |                             | .data         |       1720 |       1732 |       12 |        0.7 |
|            |                             |                             | .text         |     731232 |     731968 |      736 |        0.1 |
|            | window-app                  | BRD4161A                    | .bss          |     260416 |     260408 |       -8 |       -0.0 |
|            |                             |                             | .data         |       1724 |       1736 |       12 |        0.7 |
|            |                             |                             | .text         |     732632 |     733384 |      752 |        0.1 |
|            | lighting-app                | BRD4161A+rpc                | .bss          |     260296 |     260288 |       -8 |       -0.0 |
|            |                             |                             | .data         |       1844 |       1856 |       12 |        0.7 |
|            |                             |                             | .text         |     730136 |     730872 |      736 |        0.1 |
| esp32      | all-clusters-app            | c3devkit                    | .dram0.bss    |      65336 |      65384 |       48 |        0.1 |
|            |                             |                             | .dram0.data   |      16280 |      16288 |        8 |        0.0 |
|            |                             |                             | .flash.rodata |     196856 |     196960 |      104 |        0.1 |
|            |                             |                             | .flash.text   |     857648 |     857772 |      124 |        0.0 |
|            |                             |                             | .iram0.text   |      57320 |      57320 |        0 |        0.0 |
|            | bridge-app                  | default                     | .dram0.bss    |      65648 |      65664 |       16 |        0.0 |
|            |                             |                             | .dram0.data   |      19980 |      19988 |        8 |        0.0 |
|            |                             |                             | .flash.rodata |     156316 |     156392 |       76 |        0.0 |
|            |                             |                             | .flash.text   |     749327 |     749507 |      180 |        0.0 |
|            |                             |                             | .iram0.text   |     115755 |     115755 |        0 |        0.0 |
|            | ipv6only-app                | default                     | .dram0.bss    |      19760 |      19760 |        0 |        0.0 |
|            |                             |                             | .dram0.data   |      16232 |      16232 |        0 |        0.0 |
|            |                             |                             | .flash.rodata |      99040 |      99040 |        0 |        0.0 |
|            |                             |                             | .flash.text   |     471015 |     471015 |        0 |        0.0 |
|            |                             |                             | .iram0.text   |      84275 |      84275 |        0 |        0.0 |
|            | lock-app                    | default                     | .dram0.bss    |      47384 |      47392 |        8 |        0.0 |
|            |                             |                             | .dram0.data   |      19796 |      19804 |        8 |        0.0 |
|            |                             |                             | .flash.rodata |     157284 |     157368 |       84 |        0.1 |
|            |                             |                             | .flash.text   |     751871 |     752099 |      228 |        0.0 |
|            |                             |                             | .iram0.text   |     115755 |     115755 |        0 |        0.0 |
|            | persistent-storage          | default                     | .dram0.bss    |       2152 |       2152 |        0 |        0.0 |
|            |                             |                             | .dram0.data   |      10128 |      10128 |        0 |        0.0 |
|            |                             |                             | .flash.rodata |      30896 |      30896 |        0 |        0.0 |
|            |                             |                             | .flash.text   |      97451 |      97451 |        0 |        0.0 |
|            |                             |                             | .iram0.text   |      41151 |      41151 |        0 |        0.0 |
|            | pigweed-app                 | default                     | .dram0.bss    |       3456 |       3456 |        0 |        0.0 |
|            |                             |                             | .dram0.data   |      10208 |      10208 |        0 |        0.0 |
|            |                             |                             | .flash.rodata |      35440 |      35440 |        0 |        0.0 |
|            |                             |                             | .flash.text   |      98183 |      98183 |        0 |        0.0 |
|            |                             |                             | .iram0.text   |      41727 |      41727 |        0 |        0.0 |
|            | shell                       | default                     | .dram0.bss    |      36640 |      36640 |        0 |        0.0 |
|            |                             |                             | .dram0.data   |      19692 |      19700 |        8 |        0.0 |
|            |                             |                             | .flash.rodata |     142868 |     142936 |       68 |        0.0 |
|            |                             |                             | .flash.text   |     724799 |     724975 |      176 |        0.0 |
|            |                             |                             | .iram0.text   |     119015 |     119015 |        0 |        0.0 |
|            | all-clusters-app            | m5stack                     | .dram0.bss    |      64480 |      64536 |       56 |        0.1 |
|            |                             |                             | .dram0.data   |      32132 |      32140 |        8 |        0.0 |
|            |                             |                             | .flash.rodata |     198420 |     198536 |      116 |        0.1 |
|            |                             |                             | .flash.text   |     863859 |     864027 |      168 |        0.0 |
|            |                             |                             | .iram0.text   |     121839 |     121839 |        0 |        0.0 |
|            | temperature-measurement-app | optimize                    | .dram0.bss    |      45880 |      45888 |        8 |        0.0 |
|            |                             |                             | .dram0.data   |      15204 |      15212 |        8 |        0.1 |
|            |                             |                             | .flash.rodata |     110084 |     110160 |       76 |        0.1 |
|            |                             |                             | .flash.text   |     716447 |     716603 |      156 |        0.0 |
|            |                             |                             | .iram0.text   |     107475 |     107475 |        0 |        0.0 |
| k32w       | lock-app                    | k32w061+debug               | .bss          |      66460 |      66476 |       16 |        0.0 |
|            |                             |                             | .data         |       1820 |       1832 |       12 |        0.7 |
|            |                             |                             | .text         |     496100 |     496372 |      272 |        0.1 |
|            | shell                       | k32w061+debug               | .bss          |      56296 |      56296 |        0 |        0.0 |
|            |                             |                             | .data         |        660 |        672 |       12 |        1.8 |
|            |                             |                             | .text         |     365460 |     365700 |      240 |        0.1 |
|            | lighting-app                | k32w061+se05x+release       | .bss          |      76400 |      76424 |       24 |        0.0 |
|            |                             |                             | .data         |       1884 |       1896 |       12 |        0.6 |
|            |                             |                             | .text         |     594100 |     594384 |      284 |        0.0 |
| linux      | all-clusters-app            | debug                       | *unknown*     |       4064 |       4064 |        0 |        0.0 |
|            |                             |                             | FLASH         |    1309123 |    1297027 |   -12096 |       -0.9 |
|            |                             |                             | RAM           |     115592 |     116400 |      808 |        0.7 |
|            | chip-tool                   | debug                       | *unknown*     |       4128 |       4128 |        0 |        0.0 |
|            |                             |                             | FLASH         |    2560593 |    2545297 |   -15296 |       -0.6 |
|            |                             |                             | RAM           |      89088 |      89832 |      744 |        0.8 |
|            | ota-provider-app            | debug                       | *unknown*     |       3928 |       3928 |        0 |        0.0 |
|            |                             |                             | FLASH         |    1011854 |     999638 |   -12216 |       -1.2 |
|            |                             |                             | RAM           |      60792 |      61520 |      728 |        1.2 |
|            | shell                       | debug                       | *unknown*     |       3896 |       3896 |        0 |        0.0 |
|            |                             |                             | FLASH         |     696093 |     697133 |     1040 |        0.1 |
|            |                             |                             | RAM           |      55832 |      56576 |      744 |        1.3 |
|            | tv-app                      | debug                       | *unknown*     |       4184 |       4184 |        0 |        0.0 |
|            |                             |                             | FLASH         |    1430476 |    1417516 |   -12960 |       -0.9 |
|            |                             |                             | RAM           |     267808 |     268584 |      776 |        0.3 |
|            | bridge-app                  | debug+rpc                   | *unknown*     |       3912 |       3912 |        0 |        0.0 |
|            |                             |                             | FLASH         |    1058053 |    1045973 |   -12080 |       -1.1 |
|            |                             |                             | RAM           |      79992 |      80752 |      760 |        1.0 |
|            | lighting-app                | debug+rpc                   | *unknown*     |       4096 |       4096 |        0 |        0.0 |
|            |                             |                             | FLASH         |    1244367 |    1232239 |   -12128 |       -1.0 |
|            |                             |                             | RAM           |      94808 |      95552 |      744 |        0.8 |
| mbed       | lighting-app                | CY8CPROTO_062_4343W+release | .bss          |     171396 |     171420 |       24 |        0.0 |
|            |                             |                             | .data         |       5016 |       5024 |        8 |        0.2 |
|            |                             |                             | .heap         |     861056 |     861024 |      -32 |       -0.0 |
|            |                             |                             | .text         |    1109500 |    1110012 |      512 |        0.0 |
|            | lock-app                    | CY8CPROTO_062_4343W+release | .bss          |     170044 |     170060 |       16 |        0.0 |
|            |                             |                             | .data         |       4960 |       4976 |       16 |        0.3 |
|            |                             |                             | .heap         |     862464 |     862432 |      -32 |       -0.0 |
|            |                             |                             | .text         |    1087356 |    1087804 |      448 |        0.0 |
| nrfconnect | lighting-app                | nrf52840dk_nrf52840         | bss           |     112892 |     112908 |       16 |        0.0 |
|            |                             |                             | rodata        |  536312592 |  536312224 |     -368 |       -0.0 |
|            |                             |                             | text          |     557172 |     557536 |      364 |        0.1 |
|            | lock-app                    | nrf52840dk_nrf52840         | bss           |     111508 |     111520 |       12 |        0.0 |
|            |                             |                             | rodata        |  536330512 |  536330144 |     -368 |       -0.0 |
|            |                             |                             | text          |     539286 |     539654 |      368 |        0.1 |
|            | pigweed-app                 | nrf52840dk_nrf52840         | bss           |      52232 |      52232 |        0 |        0.0 |
|            |                             |                             | rodata        |  536533632 |  536533632 |        0 |        0.0 |
|            |                             |                             | text          |     336330 |     336330 |        0 |        0.0 |
|            | pump-app                    | nrf52840dk_nrf52840         | bss           |     111932 |     111940 |        8 |        0.0 |
|            |                             |                             | rodata        |  536327216 |  536326832 |     -384 |       -0.0 |
|            |                             |                             | text          |     542596 |     542964 |      368 |        0.1 |
|            | pump-controller-app         | nrf52840dk_nrf52840         | bss           |     111432 |     111440 |        8 |        0.0 |
|            |                             |                             | rodata        |  536331968 |  536331600 |     -368 |       -0.0 |
|            |                             |                             | text          |     537840 |     538204 |      364 |        0.1 |
|            | shell                       | nrf52840dk_nrf52840         | bss           |     110236 |     110232 |       -4 |       -0.0 |
|            |                             |                             | rodata        |  536347344 |  536346944 |     -400 |       -0.0 |
|            |                             |                             | text          |     522472 |     522874 |      402 |        0.1 |
|            | lighting-app                | nrf52840dk_nrf52840+rpc     | bss           |     109108 |     109128 |       20 |        0.0 |
|            |                             |                             | rodata        |  536336832 |  536336464 |     -368 |       -0.0 |
|            |                             |                             | text          |     533060 |     533424 |      364 |        0.1 |
|            |                             | nrf5340dk_nrf5340_cpuapp    | bss           |     115860 |     115880 |       20 |        0.0 |
|            |                             |                             | rodata        |  536374592 |  536374224 |     -368 |       -0.0 |
|            |                             |                             | text          |     494496 |     494860 |      364 |        0.1 |
|            | lock-app                    | nrf5340dk_nrf5340_cpuapp    | bss           |     114480 |     114488 |        8 |        0.0 |
|            |                             |                             | rodata        |  536392528 |  536392160 |     -368 |       -0.0 |
|            |                             |                             | text          |     476594 |     476962 |      368 |        0.1 |
|            | shell                       | nrf5340dk_nrf5340_cpuapp    | bss           |     112816 |     112812 |       -4 |       -0.0 |
|            |                             |                             | rodata        |  536420176 |  536419776 |     -400 |       -0.0 |
|            |                             |                             | text          |     448982 |     449384 |      402 |        0.1 |
| p6         | lock-app                    | default                     | .bss          |      67976 |      67992 |       16 |        0.0 |
|            |                             |                             | .data         |       2384 |       2400 |       16 |        0.7 |
|            |                             |                             | .heap         |     967080 |     967048 |      -32 |       -0.0 |
|            |                             |                             | .text         |    1101416 |    1102200 |      784 |        0.1 |
| qpg        | lighting-app                | qpg6100+debug               | .bss          |      88040 |      88032 |       -8 |       -0.0 |
|            |                             |                             | .data         |        980 |        992 |       12 |        1.2 |
|            |                             |                             | .text         |     482772 |     483056 |      284 |        0.1 |
|            | lock-app                    | qpg6100+debug               | .bss          |      88104 |      88096 |       -8 |       -0.0 |
|            |                             |                             | .data         |        916 |        928 |       12 |        1.3 |
|            |                             |                             | .text         |     459060 |     459340 |      280 |        0.1 |
|            | persistent-storage-app      | qpg6100+debug               | .bss          |      88760 |      88736 |      -24 |       -0.0 |
|            |                             |                             | .data         |        268 |        288 |       20 |        7.5 |
|            |                             |                             | .text         |     107092 |     108788 |     1696 |        1.6 |
| telink     | lighting-app                | tlsr9518adk80d              | bss           |      71404 |      71416 |       12 |        0.0 |
|            |                             |                             | noinit        |      33216 |      33216 |        0 |        0.0 |
|            |                             |                             | text          |     444114 |     444454 |      340 |        0.1 |

</details>

File: scripts/tools/memory/platform/linux.cfg (hunk 1 above)
Comment (kpschoedel) [reply scripts/tools/memory/platform/linux.cfg:64]: I think it would make sense for the reports to include both the aggregation of loaded segments (generally as read-only vs read/write) and platform-specific ‘important’ sections (e.g. .text/.data/.bss or equivalents).

File: .github/workflows/examples-efr32.yaml [hunk 2]
```diff
@@ -57,25 +60,29 @@ jobs:
                    .environment/pigweed-venv/*.log
             - name: Build example EFR32 Lock App for BRD4161A
               timeout-minutes: 10
-              run:
-                  scripts/examples/gn_efr32_example.sh examples/lock-app/efr32/
-                  out/lock_app_debug BRD4161A
+              run: |
+                  scripts/examples/gn_efr32_example.sh examples/lock-app/efr32/ out/lock_app_debug BRD4161A
+                  .environment/pigweed-venv/bin/python3 scripts/tools/memory/gh_sizes.py efr32 BRD4161A lock-app \
```
Comment (bzbarsky-apple) [inline .github/workflows/examples-efr32.yaml:65]: Is this running in a setup where we have to hardcode that path instead of just using `python3`?

File: .github/workflows/examples-linux-standalone.yaml [hunk 3]
```diff
@@ -57,36 +60,62 @@ jobs:
...
               timeout-minutes: 5
-              run:
-                  scripts/examples/gn_build_example.sh examples/chip-tool
-                  out/chip_tool_debug
+              run: |
+                  scripts/examples/gn_build_example.sh examples/chip-tool out/chip_tool_debug
+                  .environment/pigweed-venv/bin/python3 scripts/tools/memory/gh_sizes.py \
+                    linux debug chip-tool \
+                    out/chip_tool_debug/chip-tool \
+                    /tmp/bloat_reports/ --collect-method readelf --by=region
```
Comment (bzbarsky-apple) [inline .github/workflows/examples-linux-standalone.yaml:68]: Why did the other yamls not need the --collect-method bits?  Might be worth a comment.

File: scripts/examples/esp_example.sh [hunk 4]
```diff
@@ -48,4 +48,6 @@ rm -f "$root"/sdkconfig
     echo "build $sdkconfig_name failed"
     exit 1
 }
-cp "$root"/build/chip-"$app".elf "$root"/build/"${sdkconfig_name%".defaults"}"-chip-"$app".elf
+config="${sdkconfig_name%.defaults}"
+output="$root"/build/"$config"-chip-"$app".elf
+cp "$root"/build/chip-"$app".elf "$output"
```
Comment (bzbarsky-apple) [inline scripts/examples/esp_example.sh:53]: This is fine, but why is it needed here?

File: scripts/examples/gn_efr32_example.sh [hunk 5]
```diff
@@ -20,23 +20,21 @@ set -e
...
-    else
-        gn gen --check --fail-on-unused-args --root="$1" --args="efr32_board=\"$3\"" "$2/$3" "$4"
-    fi
-    ninja -v -C "$2/$3"
-    #print stats
-    arm-none-eabi-size -A "$2"/"$3"/*.out
+    board="$EFR32_BOARD"
 fi
+
+gn gen --check --fail-on-unused-args --root="$root" --args="efr32_board=\"$board\"" "$out/$board" "$@"
```
Comment (bzbarsky-apple) [inline scripts/examples/gn_efr32_example.sh:39]: This is changing the command executed when there are only 2 args (now passes `--args="efr32_board=\"$EFR32_BOARD"\""` instead of `--args=""`).  Does this lead to the same behavior?

And again, why are these changes needed?  And why are we dropping the `srm-none-eabi-size` bits?

File: scripts/tools/memory/gh_report.py [hunk 6]
```diff
@@ -0,0 +1,548 @@
...
+                    logging.info('    %s %s', parent.id, parent.name)
+
+        # Download and add required artifacts.
+        for i in required_artifact_ids:
+            logging.debug('Download artifact %d', i)
+            try:
+                blob = self.gh.actions.download_artifact(i, 'zip')
+            except Exception as e:
+                logging.error('Failed to download artifact %d: %s', i, e)
+            self.add_sizes_from_zipfile(io.BytesIO(blob), {'artifact': i})
```
Comment (bzbarsky-apple) [inline scripts/tools/memory/gh_report.py:255]: But some of the artifacts we uploaded were a single JSON file, not zipfiles, no?  See `examples-infineon.yaml`.

**Review by bzbarsky-apple — APPROVED** [review]
I did not carefully review and understand all the `scripts/tools/memory/memdf` changes....

I also did not verify the various section names under `scripts/tools/memory/platform/`

File: scripts/examples/gn_efr32_example.sh (hunk 5 above)
Comment (kpschoedel) [reply scripts/examples/gn_efr32_example.sh:39]: Another dev leftover; will remove. (I had previously had `gh_sizes` run from some of these scripts, which made `arm-none-eabi-size` redundant.)

File: scripts/tools/memory/gh_report.py (hunk 6 above)
Comment (kpschoedel) [reply scripts/tools/memory/gh_report.py:255]: Github unconditionally wraps workflows artifacts in a zip; there's no other way to get them.
